feat(calendar): add firstDayOfWeek and days abbreviations

Apercu opened the pull request:
This commit adds the ability to set the firstDayOfWeek for Europeans that often starts on monday rather than sunday, and to set a days abbreviations array for days first letter.

This part could be improved though, but don't really know how, feel free to tell me if you got an idea.

Related to #2099.

oliviertassinari commented on src/date-picker/date-picker.jsx:
```diff
@@ -43,6 +49,12 @@ const DatePicker = React.createClass({
+     * Used to change the first day of week
+     * Defaults to 0 (Sunday)
+     */
+    firstDayOfWeek: React.PropTypes.number,
```
Can't we use Intl to do this work for us?
This can help https://github.com/callemall/material-ui/pull/2282

oliviertassinari commented on src/date-picker/date-picker.jsx:
```diff
@@ -31,6 +31,12 @@ const DatePicker = React.createClass({
+     * Used to change the days abbreviation in the calendar
+     * Defaults to S M T W T F S
+     */
+    daysArray: React.PropTypes.array,
```
Same comment, using Intl would be far simpler for people using the component.

Apercu commented:
@oliviertassinari Thanks, I will update my PR and ping again

Apercu commented:
I though it would be a little overkill to create 7 dates only to find a char, don't we have any way to access the Intl locales without having to do this?

Apercu commented:
@oliviertassinari I got the Intl for the days abbreviations, but I dunno how to get the first day of the week without having to rely on the `getDate` method of the Date object, which doesn't seems to follow the current locale.

oliviertassinari commented:
> Sorry, this diff is unavailable.

:grin:

> I got the Intl for the days abbreviations

Awesome!

> but I dunno how to get the first day of the week without having to rely on the getDate method of the Date object, which doesn't seems to follow the current locale.

Yeah, I think that it's the most difficult point.

Apercu commented:
Ah sorry, I thought rebasing would be better.

Any idea why the build cannot seems to find the `Intl` variable?

oliviertassinari commented:
> Any idea why the build cannot seems to find the Intl variable?

I can't see the line changes right now.
Regarding the latest point. Basicaly, we are looking for something like: http://momentjs.com/docs/#/get-set/iso-weekday/.

Apercu commented:
Will try to fix the commit.

Hum yeah, I thought you didn't want to add an extra dependency like moment, but it would be pretty easy with it yes

oliviertassinari commented:
> I thought you didn't want to add an extra dependency like moment

I definitly don't want to. Here is their implementation https://github.com/moment/moment/blob/11b18a8c5e8a7970d38352f96f226c3351b89a43/src/lib/units/day-of-week.js#L154.

Apercu commented:
Okay the commit is now fixed, I removed the merge so you can see the diff.

Also I looked a bit this method and actually don't know how we can use it to fit our needs (my mind is really bad with dates and timezone :smile:), and do we really will have to re-implement each method we need to not include a library?

Apercu commented:
@oliviertassinari I updated the changes to handle the build issue and use a fallback when `Intl` is not available. Would you consider merge this PR with the `firstDayOfWeek` parameter?

I'm thinking the changes with our own implementation of  the first day detection would be out-of-range of this PR and might need a closer look to the possibilities

oliviertassinari commented:
> Would you consider merge this PR with the firstDayOfWeek parameter?

If we don't have better option, yes.

oliviertassinari commented on src/utils/date-time.js:
```diff
@@ -77,33 +78,52 @@ export default {
+    return new Date(now.setDate(now.getDate() - now.getDay()));
+  },
+
+  getWeekArray(d, firstDayOfWeek = 0) {
```
Can you set the default value in the `getDefaultProps` of the `DateTimePicker` component?
This will automatically be used by the documentation.

oliviertassinari commented on src/date-picker/date-picker.jsx:
```diff
@@ -43,6 +43,12 @@ const DatePicker = React.createClass({
 
     /**
+     * Used to change the first day of week
+     * Defaults to 0 (Sunday)
```
Can you remove this line? The documentation use `getDefaultProps` to display it automatically.

oliviertassinari commented on src/date-picker/calendar.jsx:
```diff
@@ -325,13 +327,11 @@ const Calendar = React.createClass({
-              <li style={weekTitleDayStyle}>T</li>
-              <li style={weekTitleDayStyle}>F</li>
-              <li style={weekTitleDayStyle}>S</li>
+              {[...Array(7)].map((e, i) => (
```
Could we move `[...Array(7)` out of the render method?
This will save memory allocation between two render calls.

Apercu commented:
@oliviertassinari I changed the things you pointed out ;)

Apercu commented:
Wait I need to change something

Apercu commented:
@oliviertassinari All good :ok_hand: I didn't worked when you didn't provided a `DateTimeFormat`.

Sorry for all the pings :confused:

oliviertassinari commented on src/date-picker/calendar.jsx:
```diff
@@ -59,6 +60,7 @@ const Calendar = React.createClass({
       selectedDate: this.props.initialDate,
       transitionDirection: 'left',
       transitionEnter: true,
+      daysArray: [...Array(7)],
```
This will be allocation for each instance of the component. Why not simply move outside of the component?

Apercu replied:
Like a global variable? Where do you want to put it exactly?

oliviertassinari replied:
Yes, you can do something like:
```js
....
import ThemeManager from '../styles/theme-manager';
import DefaultRawTheme from '../styles/raw-themes/light-raw-theme';

const daysArray = [...Array(7)];

const Calendar = React.createClass({
...
```

oliviertassinari commented:
Could you add `firstDayOfWeek={1}` to the `DatePickerExampleInternational` example?
That we improve our current `french` example.

oliviertassinari commented on src/date-picker/date-picker.jsx:
```diff
@@ -43,6 +43,11 @@ const DatePicker = React.createClass({
     /**
+     * Used to change the first day of week.
+     */
+    firstDayOfWeek: React.PropTypes.number,
```
I think it would be good to add something link:
`The first day of the week drastically varies between Saturday, Sunday, or Monday (or even Friday) between different locales`

Apercu commented:
@oliviertassinari I changed the docs, comments and the `daysArray` variable.

Nice catch for ecma the issue, really interesting

oliviertassinari commented:
@Apercu Really nice PR :+1:
@alitaheri @newoga Do you want to do a final review before merging?

mbrookes commented:
Please could you note the use of the `firstDayOfWeek` property in `docs/src/app/components/pages/components/DatePicker/Page.jsx` in `descriptions.localised`?

alitaheri commented on src/date-picker/date-picker.jsx:
```diff
@@ -43,6 +43,12 @@ const DatePicker = React.createClass({
+     * Used to change the first day of week. It drastically varies from
+     * Saturday to Monday (could even be Friday) between different locales.
+     */
+    firstDayOfWeek: React.PropTypes.number,
```
Might not be a bad idea to clarify what the default value, `0` means Sunday.

oliviertassinari replied:
I haven't though about this. I agree :grin:

Apercu replied:
Grrrr :grin: that's what I did in my first commit! :smile:

oliviertassinari replied:
Sorry :dizzy: !

Apercu replied:
Haha no problem I'm kidding

alitaheri commented:
Great work @Apercu, That one thing is my only concern. In my country the first day of week is Saturday. Although I know how america works. But I had to check the source to be sure. So a simple clarification can save the users a bit of time. Thanks for the awesome work :+1: :+1:

mbrookes commented:
@alitaheri - that's an interesting point. IMHO without localisation applied, I believe Material-UI date handling should be "international" by default. There's an open issue to make the default date displayed follow [ISO8601] (https://en.wikipedia.org/wiki/ISO_8601) (YYYY-MM-DD), as the US format is ambiguous everywhere, including in the US!

By the same token, I think the default the first day of the week should be also follow ISO8601. This happens to be Monday for better or worse, but at least it would be consistent with a standard, rather than following one countries cultural norm.

oliviertassinari commented:
> By the same token, I think the default the first day of the week should be also follow ISO8601. This happens to be Monday for better or worse, but at least it would be consistent with a standard

I agree, but we have to wait for the next breaking changes release.

mbrookes commented:
In which case, should `firstDayOfWeek: 0` equal Monday?

alitaheri commented:
@mbrookes I completely agree. Following the standard is the best possible way to tackle this. But as @oliviertassinari said, it will have to wait since it's a breaking change. so we need to find a way to ease into it or find a way to warn the users ahead of time if transition is not possible.

mbrookes commented:
Yeah, don't want to stand in the way of this PR. Great work BTW @Apercu!

newoga commented:
@alitaheri @oliviertassinari Sorry, I've been absolutely slammed.If you both are satisfied, feel free to merge. Seems like there might be some topics to revisit but if the code quality is solid, I'm happy :smile: 

I'll review the changes either way tonight

Apercu commented:
I updated the commit to match the review of  @alitaheri.

For the point raised by @mbrookes I totally agree, but do you think we will remember to change this in the next breaking version?

oliviertassinari commented:
@Apercu Thanks :tada: !

> but do you think we will remember to change this in the next breaking version

I have added it here: https://github.com/callemall/material-ui/issues/2109.

Apercu commented:
@oliviertassinari @alitaheri @mbrookes @newoga Thanks for the positive feedbacks and your work! Happy to contribute :tada: 

(Should be the last ping from me (on this PR!) :smile:)

alitaheri commented:
Thanks a lot, this was surely an important feature to have! :+1: :tada: :tada:
